File: myulib-core/src/main/java/com/myudog/myulib/internal/scheduler/EffectTicker.java
package com.myudog.myulib.internal.scheduler;
import net.fabricmc.fabric.api.event.lifecycle.v1.ServerTickEvents;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Supplier;
public final class EffectTicker {
    private static final CopyOnWriteArrayList<Supplier<Boolean>> activeTasks = new CopyOnWriteArrayList<>();
    private static final AtomicBoolean registered = new AtomicBoolean(false);
    private EffectTicker() {
    }
    public static void register() {
        if (registered.compareAndSet(false, true)) {
            ServerTickEvents.START_SERVER_TICK.register(server -> tick());
        }
    }
    public static void addTask(Supplier<Boolean> task) {
        register();
        activeTasks.add(task);
    }
    private static void tick() {
        for (Supplier<Boolean> task : activeTasks) {
            boolean keepGoing;
            try {
                keepGoing = Boolean.TRUE.equals(task.get());
            } catch (Throwable throwable) {
                keepGoing = false;
            }
            if (!keepGoing) {
                activeTasks.remove(task);
            }
        }
    }
}

File: src/main/kotlin/com/myudog/myulib/internal/scheduler/EffectTicker.kt
package com.myudog.myulib.internal.scheduler

import net.fabricmc.fabric.api.event.lifecycle.v1.ServerTickEvents
import java.util.concurrent.CopyOnWriteArrayList

internal object EffectTicker {
    // 儲存所有當前活動的動畫任務
    private val activeTasks = CopyOnWriteArrayList<() -> Boolean>()

    fun register() {
        ServerTickEvents.START_SERVER_TICK.register { _ ->
            val iterator = activeTasks.iterator()
            while (iterator.hasNext()) {
                val task = iterator.next()
                // 如果任務返回 false，代表已結束，將其移除
                if (!task()) activeTasks.remove(task)
            }
        }
    }

    fun addTask(task: () -> Boolean) {
        activeTasks.add(task)
    }
}